src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -0,0 +1,309 @@
+            }
+        }
+
+        /// <summary>
+        /// Returns the value of the key-value pair added to the activity with 'WithBaggage'.
+        /// Returns null if that key does not exist.  
+        /// </summary>
+        public string GetBaggageItem(string key)
+        {
+            foreach (var keyValue in Baggage)
+                if (key == keyValue.Key)
+                    return keyValue.Value;
+            return null;
+        }
+
+        /* Constructors  Builder methods */
+
+        /// <summary>
+        /// Note that Activity has a 'builder' pattern, where you call the constructor, a number of 'With*' APIs and then
+        /// call 'Activity.Start' to build the activity.   You MUST call Start before using it 
+        /// </summary>
+        /// <param name="operationName">Operations name <see cref="OperationName"/></param>
+        public Activity(string operationName)
+        {
+            OperationName = operationName;
+        }
+
+        /// <summary>
+        /// Update the Activity to have a tag with an additional 'key' and value 'value'.
+        /// This shows up in the 'Tags' eumeration.   It is meant for information that
+        /// is useful to log but not needed for runtime control (for the latter, use Baggage)
+        /// </summary>
+        /// <returns>'this' for convinient chaining</returns>
+        public Activity AddTag(string key, string value)
+        {
+            _tags = new KeyValueListNode() { keyValue = new KeyValuePair<string, string>(key, value), Next = _tags };
+            return this;
+        }
+
+        /// <summary>
+        /// Update the Activity to have baggage with an additional 'key' and value 'value'.
+        /// This shows up in the 'Baggage' eumeration as well as the 'GetBaggageItem' API.
+        /// Baggage is mean for information that is needed for runtime control.   For information 
+        /// that is simply useful to show up in the log with the activity use Tags.   
+        /// Returns 'this' for convinient chaining.
+        /// </summary>
+        /// <returns>'this' for convinient chaining</returns>
+        public Activity AddBaggage(string key, string value)
+        {
+            _baggage = new KeyValueListNode() { keyValue = new KeyValuePair<string, string>(key, value), Next = _baggage };
+            return this;
+        }
+
+        /// <summary>
+        /// Updates the Activity To indicate that the activity with ID 'parentID' 
+        /// caused this activity.   This is only intended to be used at 'boundary' 
+        /// scenarios where an activity from another process loggically started 
+        /// this activity. The Parent ID shows up the Tags (as well as the ParentID 
+        /// property), and can be used to reconstruct the causal tree.  
+        /// Returns 'this' for convinient chaining.
+        /// </summary>
+        public Activity SetParentId(string parentId)
+        {
+            ParentId = parentId;
```

Comment:
ParentId without Parent property means that parent activity is external. 
This is a good catch though: if parent is already set, but SetParentId is called later, it may lead to inconsistent state. I'll fix that